plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,290 @@
+
+import type { Request } from 'express';
+
+import type { KubernetesRequestAuth } from '@backstage/plugin-kubernetes-common';
+
+import { ClusterDetails, KubernetesClustersSupplier } from '../types/types';
+import { KubernetesAuthTranslatorGenerator } from '../kubernetes-auth-translator/KubernetesAuthTranslatorGenerator';
+
+const HEADER_KUBERNETES_CLUSTERS: string = 'X-Kubernetes-Clusters';
+
+const ERROR_BAD_REQUEST: number = 400;
+const ERROR_INTERNAL_SERVER: number = 500;
+
+const TRANSLATOR_TYPE_TOKEN: string = 'token';
+const TRANSLATOR_TYPE_NOOP: string = 'serviceAccount';
+
+export interface KubernetesProxyServices {
+  kcs: KubernetesClustersSupplier;
+}
+
+export interface KubernetesProxyResponse {
+  code: number;
+  data: any;
+  cluster?: string;
+}
+
+interface KubernetesProxyClusters {
+  [key: string]: string;
+}
+
+export class KubernetesProxy {
+  constructor(protected readonly logger: Logger) {}
+
+  public async handleProxyRequest(
+    services: KubernetesProxyServices,
+    req: Request,
+  ): Promise<any> {
+    const krc = this.getKubernetesRequestedClusters(req);
+    const details = await this.getClusterDetails(services.kcs, krc);
+
+    const responses = await Promise.all(
+      details.map(async d => {
+        const response = await this.makeRequestToCluster(d, req);
+        return response;
+      }),
+    );
+
+    const data: { [key: string]: any } = {};
+    const codes: number[] = [];
+
+    responses.forEach(kpr => {
+      if (kpr.cluster) {
+        data[kpr.cluster] = kpr.data;
+        codes.push(kpr.code);
+      }
+    });
+
+    const code = this.getBestResponseCode(codes);
+
+    const res: KubernetesProxyResponse = {
+      code,
+      data,
+    };
+
+    return res;
+  }
+
+  protected getKubernetesRequestedClusters(
+    req: Request,
+  ): KubernetesProxyClusters {
+    const encodedClusters: string =
+      req.header(HEADER_KUBERNETES_CLUSTERS) ?? '';
+
+    if (!encodedClusters) {
+      return {};
+    }
+
+    try {
+      const decodedClusters = Buffer.from(encodedClusters, 'base64').toString();
+      const clusters: KubernetesProxyClusters = JSON.parse(decodedClusters);
+      return clusters;
+    } catch (e: any) {
+      this.logger.error(
+        `error with encoded cluster header: ${JSON.stringify(e)}`,
+      );
+    }
+    return {};
+  }
+
+  protected async getClusterDetails(
+    clusterSupplier: KubernetesClustersSupplier,
+    krc: KubernetesProxyClusters,
+  ): Promise<ClusterDetails[]> {
+    const clusters = await clusterSupplier.getClusters();
+
+    const clusterNames = Object.keys(krc);
+
+    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));
+
+    const clusterDetailsAuth = await Promise.all(
+      clusterDetails.map(async c => {
+        const authMetadata: KubernetesRequestAuth = {
+          token: krc[c.name],
+        };
+        const cAuth = await this.configureAuth(c, authMetadata);
+        return cAuth;
+      }),
+    );
```

Comment:
If we're going to take @chinigo's [advice about auth](https://github.com/backstage/backstage/issues/12231#issuecomment-1192875578):

> The proxy would expect calls to supply an opaque `Authorization` header, and it will blindly send them along in its fanned-out requests. Operators who want to support multi-cluster operations can use their platforms to respect that shared token at their discretion

I think there may be no need to introduce the `TokenKubernetesAuthTranslator` -- instead of using a KubernetesAuthTranslator at all, this code could enrich the clusterDetails with the token directly, like this (maybe with some copying instead of mutation in the `map`, i'm rusty):
```suggestion
    const clusterNames = Object.keys(krc);

    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));

    const clusterDetailsAuth = await Promise.all(
      clusterDetails.map(async c => {
        c.serviceAccountToken = krc[c.name];
        return c;
      }),
    );
```

Reply:
I might have gone a bit far here -- I think there could still be a place for a `configureAuth` method, but it probably doesn't need to invoke any KubernetesAuthTranslators in this version.

Reply:
I originally added this logic with the auth translator so that we could fallback if the token isn't included in the header; later on, if there isn't a token, we can use the generic auth translator for the cluster to attempt to authenticate.

Not sure if we need the fallback functionality any more, but if we want to keep it, then I would lean towards going through the auth translator for this.

Reply:
I believe that @chinigo's "opaque header" idea requires there to be _no_ fallback functionality, as it represents a pure security risk until we've thought carefully about other ways of securing this endpoint.